framework/src/functions/PiecewiseBase.C

```diff
@@ -0,0 +1,84 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PiecewiseBase.h"
+
+template <>
+InputParameters
+validParams<PiecewiseBase>()
+{
+  InputParameters params = validParams<Function>();
+
+  MooseEnum axis("x=0 y=1 z=2 0=3 1=4 2=5");
+  axis.deprecate("0", "x");
+  axis.deprecate("1", "y");
+  axis.deprecate("2", "z");
+  params.addParam<MooseEnum>(
+      "axis", axis, "The axis used (x, y, or z) if this is to be a function of position");
+  return params;
+}
+
+PiecewiseBase::PiecewiseBase(const InputParameters & parameters)
+  : Function(parameters), _has_axis(false)
+{
+}
+
+void
+PiecewiseBase::setData(const std::vector<Real> & x, const std::vector<Real> & y)
+{
+  // Size mismatch error
+  if (x.size() != y.size())
+    mooseError("In PiecewiseBase ", _name, ": Lengths of x and y data do not match.");
+
+  try
+  {
+    _linear_interp = libmesh_make_unique<LinearInterpolation>(x, y);
+  }
+  catch (std::domain_error & e)
+  {
+    mooseError("In PiecewiseBase ", _name, ": ", e.what());
+  }
+
+  if (isParamValid("axis"))
+  {
+    const MooseEnum & axis = getParam<MooseEnum>("axis");
+    switch (axis)
+    {
+      case 0:
+      case 1:
+      case 2:
+        _axis = axis;
+        break;
+      case 3:
+      case 4:
+      case 5:
+        _axis = axis - 3;
+        break;
+    }
+    _has_axis = true;
+  }
+}
```

Comment:
No need to do all this work. Since you are controlling the MooseEnum values you can just implicitly convert the enum value to an `int` (e.g. this whole case statement goes away).

Reply:
I realize you didn't write this code so you don't have to change it. I'll leave that up to you.

Reply:
Can I move setData to initialSetup? This will ensure that _linear_interp is set before running and not require inherited models to run it in their constructor, but I'm not sure if this screws up the parser ordering.

Reply:
I can move the _axis determination to the PiecewiseBase constructor, so that'll be set.